### ISSUE TITLE
attempting to check if all 512 indexes on a memory bank at once are equal to 0 with a one-liner crashes mindcode

### ISSUE THREAD

**Issue by 50275** [cardillan/mindcode#189]
i am not joking this is real code
```print(cell1[0] == cell1[1] && cell1[1] == cell1[2] && cell1[2] == cell1[3] && cell1[3] == cell1[4] && cell1[4] == cell1[5] && cell1[5] == cell1[6] && cell1[6] == cell1[7] && cell1[7] == cell1[8] && cell1[8] == cell1[9] && cell1[9] == cell1[10] && cell1[10] == cell1[11] && cell1[11] == cell1[12] && cell1[0] == cell1[1] && cell1[1] == cell1[2] && cell1[2] == cell1[3] && cell1[3] == cell1[4] && cell1[4] == cell1[5] && cell1[5] == cell1[6] && cell1[6] == cell1[7] && cell1[7] == cell1[8] && cell1[8] == cell1[9] && cell1[9] == cell1[10] && cell1[10] == cell1[11] && cell1[11] == cell1[12] && cell1[12] == cell1[13] && cell1[13] == cell1[14] && cell1[14] == cell1[15] && cell1[15] == cell1[16] && cell1[16] == cell1[17] && cell1[17] == cell1[18] && cell1[18] == cell1[19] && cell1[19] == cell1[20] && cell1[20] == cell1[21] && cell1[21] == cell1[22] && cell1[22] == cell1[23] && cell1[23] == cell1[24] && cell1[24] == cell1[25] && cell1[25] == cell1[26] && cell1[26] == cell1[27] && cell1[27] == cell1[28] && cell1[28] == cell1[29] && cell1[29] == cell1[30] && cell1[30] == cell1[31] && cell1[31] == cell1[32] && cell1[32] == cell1[33] && cell1[33] == cell1[34] && cell1[34] == cell1[35] && cell1[35] == cell1[36] && cell1[36] == cell1[37] && cell1[37] == cell1[38] && cell1[38] == cell1[39] && cell1[39] == cell1[40] && cell1[40] == cell1[41] && cell1[41] == cell1[42] && cell1[42] == cell1[43] && cell1[43] == cell1[44] && cell1[44] == cell1[45] && cell1[45] == cell1[46] && cell1[46] == cell1[47] && cell1[47] == cell1[48] && cell1[48] == cell1[49] && cell1[49] == cell1[50] && cell1[50] == cell1[51] && cell1[51] == cell1[52] && cell1[52] == cell1[53] && cell1[53] == cell1[54] && cell1[54] == cell1[55] && cell1[55] == cell1[56] && cell1[56] == cell1[57] && cell1[57] == cell1[58] && cell1[58] == cell1[59] && cell1[59] == cell1[60] && cell1[60] == cell1[61] && cell1[61] == cell1[62] && cell1[62] == cell1[63] && cell1[63] == cell1[64] && cell1[64] == cell1[65] && cell1[65] == cell1[66] && cell1[66] == cell1[67] && cell1[67] == cell1[68] && cell1[68] == cell1[69] && cell1[69] == cell1[70] && cell1[70] == cell1[71] && cell1[71] == cell1[72] && cell1[72] == cell1[73] && cell1[73] == cell1[74] && cell1[74] == cell1[75] && cell1[75] == cell1[76] && cell1[76] == cell1[77] && cell1[77] == cell1[78] && cell1[78] == cell1[79] && cell1[79] == cell1[80] && cell1[80] == cell1[81] && cell1[81] == cell1[82] && cell1[82] == cell1[83] && cell1[83] == cell1[84] && cell1[84] == cell1[85] && cell1[85] == cell1[86] && cell1[86] == cell1[87] && cell1[87] == cell1[88] && cell1[88] == cell1[89] && cell1[89] == cell1[90] && cell1[90] == cell1[91] && cell1[91] == cell1[92] && cell1[92] == cell1[93] && cell1[93] == cell1[94] && cell1[94] == cell1[95] && cell1[95] == cell1[96] && cell1[96] == cell1[97] && cell1[97] == cell1[98] && cell1[98] == cell1[99] && cell1[99] == cell1[100] && cell1[100] == cell1[101] && cell1[101] == cell1[102] && cell1[102] == cell1[103] && cell1[103] == cell1[104] && cell1[104] == cell1[105] && cell1[105] == cell1[106] && cell1[106] == cell1[107] && cell1[107] == cell1[108] && cell1[108] == cell1[109] && cell1[109] == cell1[110] && cell1[110] == cell1[111] && cell1[111] == cell1[112] && cell1[112] == cell1[113] && cell1[113] == cell1[114] && cell1[114] == cell1[115] && cell1[115] == cell1[116] && cell1[116] == cell1[117] && cell1[117] == cell1[118] && cell1[118] == cell1[119] && cell1[119] == cell1[120] && cell1[120] == cell1[121] && cell1[121] == cell1[122] && cell1[122] == cell1[123] && cell1[123] == cell1[124] && cell1[124] == cell1[125] && cell1[125] == cell1[126] && cell1[126] == cell1[127] && cell1[127] == cell1[128] && cell1[128] == cell1[129] && cell1[129] == cell1[130] && cell1[130] == cell1[131] && cell1[131] == cell1[132] && cell1[132] == cell1[133] && cell1[133] == cell1[134] && cell1[134] == cell1[135] && cell1[135] == cell1[136] && cell1[136] == cell1[137] && cell1[137] == cell1[138] && cell1[138] == cell1[139] && cell1[139] == cell1[140] && cell1[140] == cell1[141] && cell1[141] == cell1[142] && cell1[142] == cell1[143] && cell1[143] == cell1[144] && cell1[144] == cell1[145] && cell1[145] == cell1[146] && cell1[146] == cell1[147] && cell1[147] == cell1[148] && cell1[148] == cell1[149] && cell1[149] == cell1[150] && cell1[150] == cell1[151] && cell1[151] == cell1[152] && cell1[152] == cell1[153] && cell1[153] == cell1[154] && cell1[154] == cell1[155] && cell1[155] == cell1[156] && cell1[156] == cell1[157] && cell1[157] == cell1[158] && cell1[158] == cell1[159] && cell1[159] == cell1[160] && cell1[160] == cell1[161] && cell1[161] == cell1[162] && cell1[162] == cell1[163] && cell1[163] == cell1[164] && cell1[164] == cell1[165] && cell1[165] == cell1[166] && cell1[166] == cell1[167] && cell1[167] == cell1[168] && cell1[168] == cell1[169] && cell1[169] == cell1[170] && cell1[170] == cell1[171] && cell1[171] == cell1[172] && cell1[172] == cell1[173] && cell1[173] == cell1[174] && cell1[174] == cell1[175] && cell1[175] == cell1[176] && cell1[176] == cell1[177] && cell1[177] == cell1[178] && cell1[178] == cell1[179] && cell1[179] == cell1[180] && cell1[180] == cell1[181] && cell1[181] == cell1[182] && cell1[182] == cell1[183] && cell1[183] == cell1[184] && cell1[184] == cell1[185] && cell1[185] == cell1[186] && cell1[186] == cell1[187] && cell1[187] == cell1[188] && cell1[188] == cell1[189] && cell1[189] == cell1[190] && cell1[190] == cell1[191] && cell1[191] == cell1[192] && cell1[192] == cell1[193] && cell1[193] == cell1[194] && cell1[194] == cell1[195] && cell1[195] == cell1[196] && cell1[196] == cell1[197] && cell1[197] == cell1[198] && cell1[198] == cell1[199] && cell1[199] == cell1[200] && cell1[200] == cell1[201] && cell1[201] == cell1[202] && cell1[202] == cell1[203] && cell1[203] == cell1[204] && cell1[204] == cell1[205] && cell1[205] == cell1[206] && cell1[206] == cell1[207] && cell1[207] == cell1[208] && cell1[208] == cell1[209] && cell1[209] == cell1[210] && cell1[210] == cell1[211] && cell1[211] == cell1[212] && cell1[212] == cell1[213] && cell1[213] == cell1[214] && cell1[214] == cell1[215] && cell1[215] == cell1[216] && cell1[216] == cell1[217] && cell1[217] == cell1[218] && cell1[218] == cell1[219] && cell1[219] == cell1[220] && cell1[220] == cell1[221] && cell1[221] == cell1[222] && cell1[222] == cell1[223] && cell1[223] == cell1[224] && cell1[224] == cell1[225] && cell1[225] == cell1[226] && cell1[226] == cell1[227] && cell1[227] == cell1[228] && cell1[228] == cell1[229] && cell1[229] == cell1[230] && cell1[230] == cell1[231] && cell1[231] == cell1[232] && cell1[232] == cell1[233] && cell1[233] == cell1[234] && cell1[234] == cell1[235] && cell1[235] == cell1[236] && cell1[236] == cell1[237] && cell1[237] == cell1[238] && cell1[238] == cell1[239] && cell1[239] == cell1[240] && cell1[240] == cell1[241] && cell1[241] == cell1[242] && cell1[242] == cell1[243] && cell1[243] == cell1[244] && cell1[244] == cell1[245] && cell1[245] == cell1[246] && cell1[246] == cell1[247] && cell1[247] == cell1[248] && cell1[248] == cell1[249] && cell1[249] == cell1[250] && cell1[250] == cell1[251] && cell1[251] == cell1[252] && cell1[252] == cell1[253] && cell1[253] == cell1[254] && cell1[254] == cell1[255] && cell1[255] == cell1[256] && cell1[256] == cell1[257] && cell1[257] == cell1[258] && cell1[258] == cell1[259] && cell1[259] == cell1[260] && cell1[260] == cell1[261] && cell1[261] == cell1[262] && cell1[262] == cell1[263] && cell1[263] == cell1[264] && cell1[264] == cell1[265] && cell1[265] == cell1[266] && cell1[266] == cell1[267] && cell1[267] == cell1[268] && cell1[268] == cell1[269] && cell1[269] == cell1[270] && cell1[270] == cell1[271] && cell1[271] == cell1[272] && cell1[272] == cell1[273] && cell1[273] == cell1[274] && cell1[274] == cell1[275] && cell1[275] == cell1[276] && cell1[276] == cell1[277] && cell1[277] == cell1[278] && cell1[278] == cell1[279] && cell1[279] == cell1[280] && cell1[280] == cell1[281] && cell1[281] == cell1[282] && cell1[282] == cell1[283] && cell1[283] == cell1[284] && cell1[284] == cell1[285] && cell1[285] == cell1[286] && cell1[286] == cell1[287] && cell1[287] == cell1[288] && cell1[288] == cell1[289] && cell1[289] == cell1[290] && cell1[290] == cell1[291] && cell1[291] == cell1[292] && cell1[292] == cell1[293] && cell1[293] == cell1[294] && cell1[294] == cell1[295] && cell1[295] == cell1[296] && cell1[296] == cell1[297] && cell1[297] == cell1[298] && cell1[298] == cell1[299] && cell1[299] == cell1[300] && cell1[300] == cell1[301] && cell1[301] == cell1[302] && cell1[302] == cell1[303] && cell1[303] == cell1[304] && cell1[304] == cell1[305] && cell1[305] == cell1[306] && cell1[306] == cell1[307] && cell1[307] == cell1[308] && cell1[308] == cell1[309] && cell1[309] == cell1[310] && cell1[310] == cell1[311] && cell1[311] == cell1[312] && cell1[312] == cell1[313] && cell1[313] == cell1[314] && cell1[314] == cell1[315] && cell1[315] == cell1[316] && cell1[316] == cell1[317] && cell1[317] == cell1[318] && cell1[318] == cell1[319] && cell1[319] == cell1[320] && cell1[320] == cell1[321] && cell1[321] == cell1[322] && cell1[322] == cell1[323] && cell1[323] == cell1[324] && cell1[324] == cell1[325] && cell1[325] == cell1[326] && cell1[326] == cell1[327] && cell1[327] == cell1[328] && cell1[328] == cell1[329] && cell1[329] == cell1[330] && cell1[330] == cell1[331] && cell1[331] == cell1[332] && cell1[332] == cell1[333] && cell1[333] == cell1[334] && cell1[334] == cell1[335] && cell1[335] == cell1[336] && cell1[336] == cell1[337] && cell1[337] == cell1[338] && cell1[338] == cell1[339] && cell1[339] == cell1[340] && cell1[340] == cell1[341] && cell1[341] == cell1[342] && cell1[342] == cell1[343] && cell1[343] == cell1[344] && cell1[344] == cell1[345] && cell1[345] == cell1[346] && cell1[346] == cell1[347] && cell1[347] == cell1[348] && cell1[348] == cell1[349] && cell1[349] == cell1[350] && cell1[350] == cell1[351] && cell1[351] == cell1[352] && cell1[352] == cell1[353] && cell1[353] == cell1[354] && cell1[354] == cell1[355] && cell1[355] == cell1[356] && cell1[356] == cell1[357] && cell1[357] == cell1[358] && cell1[358] == cell1[359] && cell1[359] == cell1[360] && cell1[360] == cell1[361] && cell1[361] == cell1[362] && cell1[362] == cell1[363] && cell1[363] == cell1[364] && cell1[364] == cell1[365] && cell1[365] == cell1[366] && cell1[366] == cell1[367] && cell1[367] == cell1[368] && cell1[368] == cell1[369] && cell1[369] == cell1[370] && cell1[370] == cell1[371] && cell1[371] == cell1[372] && cell1[372] == cell1[373] && cell1[373] == cell1[374] && cell1[374] == cell1[375] && cell1[375] == cell1[376] && cell1[376] == cell1[377] && cell1[377] == cell1[378] && cell1[378] == cell1[379] && cell1[379] == cell1[380] && cell1[380] == cell1[381] && cell1[381] == cell1[382] && cell1[382] == cell1[383] && cell1[383] == cell1[384] && cell1[384] == cell1[385] && cell1[385] == cell1[386] && cell1[386] == cell1[387] && cell1[387] == cell1[388] && cell1[388] == cell1[389] && cell1[389] == cell1[390] && cell1[390] == cell1[391] && cell1[391] == cell1[392] && cell1[392] == cell1[393] && cell1[393] == cell1[394] && cell1[394] == cell1[395] && cell1[395] == cell1[396] && cell1[396] == cell1[397] && cell1[397] == cell1[398] && cell1[398] == cell1[399] && cell1[399] == cell1[400] && cell1[400] == cell1[401] && cell1[401] == cell1[402] && cell1[402] == cell1[403] && cell1[403] == cell1[404] && cell1[404] == cell1[405] && cell1[405] == cell1[406] && cell1[406] == cell1[407] && cell1[407] == cell1[408] && cell1[408] == cell1[409] && cell1[409] == cell1[410] && cell1[410] == cell1[411] && cell1[411] == cell1[412] && cell1[412] == cell1[413] && cell1[413] == cell1[414] && cell1[414] == cell1[415] && cell1[415] == cell1[416] && cell1[416] == cell1[417] && cell1[417] == cell1[418] && cell1[418] == cell1[419] && cell1[419] == cell1[420] && cell1[420] == cell1[421] && cell1[421] == cell1[422] && cell1[422] == cell1[423] && cell1[423] == cell1[424] && cell1[424] == cell1[425] && cell1[425] == cell1[426] && cell1[426] == cell1[427] && cell1[427] == cell1[428] && cell1[428] == cell1[429] && cell1[429] == cell1[430] && cell1[430] == cell1[431] && cell1[431] == cell1[432] && cell1[432] == cell1[433] && cell1[433] == cell1[434] && cell1[434] == cell1[435] && cell1[435] == cell1[436] && cell1[436] == cell1[437] && cell1[437] == cell1[438] && cell1[438] == cell1[439] && cell1[439] == cell1[440] && cell1[440] == cell1[441] && cell1[441] == cell1[442] && cell1[442] == cell1[443] && cell1[443] == cell1[444] && cell1[444] == cell1[445] && cell1[445] == cell1[446] && cell1[446] == cell1[447] && cell1[447] == cell1[448] && cell1[448] == cell1[449] && cell1[449] == cell1[450] && cell1[450] == cell1[451] && cell1[451] == cell1[452] && cell1[452] == cell1[453] && cell1[453] == cell1[454] && cell1[454] == cell1[455] && cell1[455] == cell1[456] && cell1[456] == cell1[457] && cell1[457] == cell1[458] && cell1[458] == cell1[459] && cell1[459] == cell1[460] && cell1[460] == cell1[461] && cell1[461] == cell1[462] && cell1[462] == cell1[463] && cell1[463] == cell1[464] && cell1[464] == cell1[465] && cell1[465] == cell1[466] && cell1[466] == cell1[467] && cell1[467] == cell1[468] && cell1[468] == cell1[469] && cell1[469] == cell1[470] && cell1[470] == cell1[471] && cell1[471] == cell1[472] && cell1[472] == cell1[473] && cell1[473] == cell1[474] && cell1[474] == cell1[475] && cell1[475] == cell1[476] && cell1[476] == cell1[477] && cell1[477] == cell1[478] && cell1[478] == cell1[479] && cell1[479] == cell1[480] && cell1[480] == cell1[481] && cell1[481] == cell1[482] && cell1[482] == cell1[483] && cell1[483] == cell1[484] && cell1[484] == cell1[485] && cell1[485] == cell1[486] && cell1[486] == cell1[487] && cell1[487] == cell1[488] && cell1[488] == cell1[489] && cell1[489] == cell1[490] && cell1[490] == cell1[491] && cell1[491] == cell1[492] && cell1[492] == cell1[493] && cell1[493] == cell1[494] && cell1[494] == cell1[495] && cell1[495] == cell1[496] && cell1[496] == cell1[497] && cell1[497] == cell1[498] && cell1[498] == cell1[499] && cell1[499] == cell1[500] && cell1[500] == cell1[501] && cell1[501] == cell1[502] && cell1[502] == cell1[503] && cell1[503] == cell1[504] && cell1[504] == cell1[505] && cell1[505] == cell1[506] && cell1[506] == cell1[507] && cell1[507] == cell1[508] && cell1[508] == cell1[509] && cell1[509] == cell1[510] && cell1[510] == cell1[511] && cell1[511] == 0); ```

(these are crash logs) [recent.txt](https://github.com/user-attachments/files/18243868/recent.txt)

**Comment by 50275** [comment]
maybe i'll use a for loop in mindcode instead of python next time

**Comment by cardillan** [comment]
Thanks for reporting the issue! I'll try to fix it in the new version of the compiler I'm writing now.

Of course, even in the best case, the resulting code would have at least 1024 instructions (one for reading the value, one for processing it), and Mindcode would perform even worse (for starters, each memory cell would be read twice), so a loop would definitely be a better approach :) In anycase, the compiler shouldn't crash.

**Comment by cardillan** [comment]
A few notes on the issue:

- The problem is that the hosting provides a relatively small stack (or so it seems), so it blows up there. The code compiles fine in the command-line compiler.
  - The stack size is platform-dependent, but it is around 1 MB on most platforms. It might be less on 32bit Windows, for example.
  - In any case, stack size can be set using the -Xss java command-line argument.
  - In my environment, the stack size needed to compile this thing is about 470k.
- I'm not sure I'll be able to provide a fix. I could try to increase the stack size, but there will always be some limit to the depth of the expression before the stack overflows.
- Using an explicit stack on the heap instead of an implicit one would mean the limit is all available memory - probably a much higher one, but it would be a lot of work and my gut feeling is that blown up stack is better than a blown up memory in the web application.
- Switching to an explicit stack could be a lot of work, and probably quite messy. On top of that, I'd probably have to evaluate the syntax tree from the leaves up towards the root, and maybe implement several passes to evaluate constants first. Yeah, I'm not going there.
  -  Even if I did this, with an expression large enough the compilation would fail in the parser, and I wouldn't be able to fix that - it's an external library.
- From the stack trace it appears that the compiler was able to get some 200+ nodes deep into the syntax tree. That's not that bad after all.

So, to recap: I'll probably won't fix this, as it is possible to use command-line compiler (with the -Xss option if needed) to compile code of this complexity.

Some observations not related to the issue itself:

- Testing for zero is not a simple business in Mindustry (see [here](https://github.com/cardillan/mindcode/blob/main/doc/syntax/SYNTAX-2-EXPRESSIONS.markdown#comparison-precision)). Using the `isZero` function from the `math` library might be useful here. (Perhaps you're already aware, but if not, it might be interesting to you.)
- When you implement the logic using a loop, Mindcode will unroll the loop if possible. It would be possible for memory cells, which have just 64 elements.

**Comment by cardillan** [comment]
Unfortunately, in the new version of the compiler the problem got worse, not better - the stack size needed to compile this code is circa 570k. The new compiler is better organized and AST tree walking is separated into several methods, which results in more method calls per node.

Still, I won't be fixing that. Expressions large enough to blow up the stack in the web app will result into hundreds on instructions, which usually isn't practical, given Mindustry Logic 1000 instruction limit.

(Generally, I'd suggest using loops instead of expressions like this wherever possible. Loops get unrolled when there's enough space, but, more importantly, work even without unrolling when there isn't.)
